# Data Engineer Challenge
​
## Descripción General
Bienvenido al desafío para Ingenieros de Datos. En esta ocasión, tendrás la oportunidad de acercarte a parte de la realidad del rol, demostrar tus habilidades y conocimientos en procesamiento de datos con python y diferentes estructuras de datos.

## Challenge
En el [archivo](https://drive.google.com/file/d/1ig2ngoXFTxP5Pa8muXo02mDTFexZzsis/view?usp=sharing) encontrarás un conjunto aproximado de 398MBs. Se pide resolver los siguientes problemas implementando funciones, usando **2 enfoques por cada problema**: Uno en el que se optimice el tiempo de ejecución, y otro en que se optimice la memoria en uso.

Tu desafío debe tener al menos 6 archivos python en la carpeta `src`. Cada uno de estos archivos correspondiente a la función del mismo nombre, con el mismo formato que se indica en las instrucciones de más abajo. Solo deja la función. Además de eso, debes tener un archivo `.ipynb` donde expliques con mayor claridad tu código. En este jupyter notebook puedes ejecutar tus funciones, medir el tiempo de ejecución, memoria en uso y explayarte según estimes conveniente. Te recomendamos fuertemente que utilices celdas markdown para que expliques el paso a paso de tu código.

**NOTA:** los archivos `.py` y `.ipynb` de interés ya están creados en la estructura del desafío, solo debes completarlos con tu solución y/o agregar los archivos que estimes convenientes.

##Instalar Dependencias

In [2]:
#instalar de forma explicita
!pip install memory-profiler==0.61.0 pyspark==3.5.3 polars==1.9.0 emoji==2.14.0 findspark==2.0.1

In [5]:
#instalar mediante un archivo
!pip install -r requirements.txt

In [7]:
#descargar los datos
!curl "https://drive.usercontent.google.com/download?id=1ig2ngoXFTxP5Pa8muXo02mDTFexZzsis&confirm=xxx" --create-dirs -o "data/tweets.json.zip"

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 57.5M  100 57.5M    0     0  16.7M      0  0:00:03  0:00:03 --:--:-- 16.7M


In [8]:
#descomprimir los datos
!unzip -n data/tweets.json.zip

Archive:  data/tweets.json.zip
  inflating: farmers-protest-tweets-2021-2-4.json  


In [9]:
file_path = "farmers-protest-tweets-2021-2-4.json"


​
#Primer Challenge Q1
1. Las top 10 fechas donde hay más tweets. Mencionar el usuario (username) que más publicaciones tiene por cada uno de esos días. Debe incluir las siguientes funciones:
```python
def q1_time(file_path: str) -> List[Tuple[datetime.date, str]]:
```
```python
def q1_memory(file_path: str) -> List[Tuple[datetime.date, str]]:
```
```python
Returns:
[(datetime.date(1999, 11, 15), "LATAM321"), (datetime.date(1999, 7, 15), "LATAM_CHI"), ...]
```

#Segundo Challenge Q2
​
2. Los top 10 emojis más usados con su respectivo conteo. Debe incluir las siguientes funciones:
```python
def q2_time(file_path: str) -> List[Tuple[str, int]]:
```
```python
def q2_memory(file_path: str) -> List[Tuple[str, int]]:
```
```python
Returns:
[("✈️", 6856), ("❤️", 5876), ...]
```

#Q3 Tercer Challenge Q3
3. El top 10 histórico de usuarios (username) más influyentes en función del conteo de las menciones (@) que registra cada uno de ellos. Debe incluir las siguientes funciones:
```python
def q3_time(file_path: str) -> List[Tuple[str, int]]:
```
```python
def q3_memory(file_path: str) -> List[Tuple[str, int]]:
```
```python
Returns:
[("LATAM321", 387), ("LATAM_CHI", 129), ...]
```

#Implementaciones Optimizadas en Memoria

In [13]:
import q1_memory as q1
from memory_profiler import memory_usage
import imp
imp.reload(q1)

mem_usage_init = memory_usage()[0]
q1.q1_memory(file_path)
mem_usage_end = memory_usage()[0]
print(f"Uso de Memoria Ram: {mem_usage_end - mem_usage_init} MiB")

Uso de Memoria Ram: 3.08984375 MiB


In [15]:
import q2_memory as q2
imp.reload(q2)

mem_usage_init = memory_usage()[0]
q2.q2_memory(file_path)
mem_usage_end = memory_usage()[0]
print(f"Uso de Memoria Ram: {mem_usage_end - mem_usage_init} MiB")

Uso de Memoria Ram: 0.515625 MiB


In [16]:
import q3_memory as q3
imp.reload(q3)

mem_usage_init = memory_usage()[0]
q3.q3_memory(file_path)
mem_usage_end = memory_usage()[0]
print(f"Uso de Memoria Ram: {mem_usage_end - mem_usage_init} MiB")

Uso de Memoria Ram: 1.2890625 MiB


##Mediciones de Memoria Ram en los tres desafios:

- **Q1** -> Uso de Memoria: 3.089 MiB
- **Q2** -> Uso de Memoria: 0.515 MiB
- **Q3** -> Uso de Memoria: 1.28 MiB

#Implementacion de Optimizaciones en Tiempo

In [18]:
import q1_time as q1
import time
imp.reload(q1)

time_init = time.time()
q1.q1_time(file_path)
time_end = time.time()
print(f"Tiempo de ejecución: {time_end - time_init:.2f} segundos")

Tiempo de ejecución: 10.12 segundos


In [19]:
import q2_time as q2
import time
imp.reload(q2)

time_init = time.time()
q2.q2_time(file_path)
time_end = time.time()
print(f"Tiempo de ejecución: {time_end - time_init:.2f} segundos")

Tiempo de ejecución: 32.80 segundos


In [20]:
import q3_time as q3
import time
imp.reload(q3)

time_init = time.time()
q3.q3_time(file_path)
time_end = time.time()
print(f"Tiempo de ejecución: {time_end - time_init:.2f} segundos")

Tiempo de ejecución: 6.77 segundos


##Mediciones de tiempo de las implementaciones:

- **Q1** -> Tiempo: 10.12 s
- **Q2** -> Tiempo: 32.80 s
- **Q3** -> Tiempo: 6.77 s